File: diagscope-test-fixtures/src/main/resources/fixtures/java-rule-parity/src/main/java/example/java/parity/Contracts.java
package example.java.parity;

import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Function;

@interface RestController {}
@interface GetMapping { String value(); }
@interface Scheduled { String cron() default ""; }
@interface Retryable {}
@interface Recover {}
@interface Timed { String value() default ""; }
@interface Service {}
@interface Repository {}
@interface KafkaListener { String[] topics(); }
@interface KafkaHandler {}
@interface Transactional { Propagation propagation() default Propagation.REQUIRED; }

enum Propagation { REQUIRED, REQUIRES_NEW, MANDATORY }

interface Logger {
    void info(String message, Object... arguments);
    void warn(String message, Object... arguments);
    void error(String message, Object... arguments);
}

interface Completion {
    Completion whenComplete(BiConsumer<Object, Throwable> callback);
}

interface ExecutorService { Completion submit(Runnable task); }
interface KafkaTemplate { Completion send(String topic, String payload); }
interface Acknowledgment { void acknowledge(); }

interface RemoteResponse {
    RemoteResponse onErrorReturn(String value);
    RemoteResponse onErrorResume(Function<Throwable, String> handler);
}

interface RemoteClient {
    void refresh();
    RemoteResponse remote();
    String describeFailure(Throwable failure);
}

interface DataSource { Connection getConnection(); }
interface Connection extends AutoCloseable {
    PreparedStatement prepareStatement(String sql);
    void close();
}
interface PreparedStatement extends AutoCloseable {
    ResultSet executeQuery();
    void close();
}
interface ResultSet extends AutoCloseable { void close(); }
interface EntityManagerFactory { EntityManager createEntityManager(); }
interface EntityManager extends AutoCloseable {
    Object find(Object type, String id);
    void close();
}
interface JdbcTemplate { DataSource getDataSource(); }

final class MDC {
    static void put(String key, String value) {}
    static void remove(String key) {}
    static void clear() {}
    static Map<String, String> getCopyOfContextMap() { return Map.of(); }
    static void setContextMap(Map<String, String> context) {}
}

File: diagscope-test-fixtures/src/main/resources/fixtures/kotlin-rule-parity/src/main/kotlin/example/kotlin/parity/Contracts.kt
package example.kotlin.parity

annotation class RestController
annotation class GetMapping(val value: String)
annotation class Scheduled(val cron: String = "")
annotation class Retryable
annotation class Recover
annotation class Timed(val value: String = "")
annotation class Service
annotation class Repository
annotation class KafkaListener(val topics: Array<String>)
annotation class KafkaHandler
annotation class Transactional(val propagation: Propagation = Propagation.REQUIRED)

enum class Propagation {
    REQUIRED,
    REQUIRES_NEW,
    MANDATORY
}

interface Logger {
    fun info(message: String, vararg arguments: Any?)
    fun warn(message: String, vararg arguments: Any?)
    fun error(message: String, vararg arguments: Any?)
}

interface Completion {
    fun whenComplete(callback: (Any?, Throwable?) -> Unit): Completion
}

interface ExecutorService {
    fun submit(task: () -> Unit): Completion
}

interface KafkaTemplate {
    fun send(topic: String, payload: String): Completion
}

interface Acknowledgment {
    fun acknowledge()
}

interface RemoteResponse {
    fun onErrorReturn(value: String): RemoteResponse
    fun onErrorResume(handler: (Throwable) -> String): RemoteResponse
}

interface RemoteClient {
    fun refresh()
    fun remote(): RemoteResponse
    fun describeFailure(failure: Throwable): String
}

interface DataSource {
    fun getConnection(): Connection
}

interface Connection : AutoCloseable {
    fun prepareStatement(sql: String): PreparedStatement
}

interface PreparedStatement : AutoCloseable {
    fun executeQuery(): ResultSet
}

interface ResultSet : AutoCloseable

interface EntityManagerFactory {
    fun createEntityManager(): EntityManager
}

interface EntityManager : AutoCloseable {
    fun find(type: Any, id: String): Any?
}

interface JdbcTemplate {
    fun getDataSource(): DataSource
}

object MDC {
    fun put(key: String, value: String) {}
    fun remove(key: String) {}
    fun clear() {}
    fun getCopyOfContextMap(): Map<String, String> = emptyMap()
    fun setContextMap(context: Map<String, String>) {}
}
